GoogleDataLogger/GoogleDataLogger/Classes/Public/GDLLogBackend.h

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+/** A convenient typedef to define the block to be called upon completion of an upload to the
+ * backend.
+ */
+typedef void (^GDLBackendCompletionBlock)(NSSet<NSURL *> *successfulUploads,
+                                          NSSet<NSURL *> *unsuccessfulUploads);
+
+/** This protocol defines the common interface for logging backend implementations. */
+@protocol GDLLogBackend <NSObject>
+
+@required
+
+/** Uploads logs to the backend using this specific backend's chosen format.
+ *
+ * @param logFiles The set of log files to upload.
+ * @param onComplete A block to invoke upon completing the upload. Has two arguments:
+ *   - successfulUploads: The set of filenames uploaded successfully.
+ *   - unsuccessfulUploads: The set of filenames not uploaded successfully.
+ */
+- (void)uploadLogs:(NSSet<NSURL *> *)logFiles onComplete:(GDLBackendCompletionBlock)onComplete;
```

Comment:
Do we want to pass through the error here as well, or only the list of uploads that failed? I can imagine wanting to check the error and handle things accordingly (in case a retry is necessary).

Reply:
Probably only the list of logs that weren't uploaded successfully. Since the individual backends do their own sending of data and error handling, the storage system really just needs to know what logs it can remove from the device, or which ones it should keep around for future sending.